File: earning-our-place-ecopol-2020/analysis/replicate_table1.py
"""
Replicate Table 1 of Bower-Bir (2021), "Earning our place, more or less."
Economia Politica 38(1): 131-170. https://doi.org/10.1007/s40888-020-00201-9

Reproduces all six published models (1a, 2a, 3a, 1b, 2b, 3b) to three decimals:
every coefficient, every standard error, every N, and the R-squared row.

Requires: pandas, numpy, statsmodels, scipy.
Run from this directory:  python3 replicate_table1.py

--------------------------------------------------------------------------------
FOUR THINGS THAT WILL SILENTLY BREAK A PORT OF THIS TABLE
--------------------------------------------------------------------------------

1. THE DEPENDENT VARIABLE IS RESCALED IN THE DATA.
   Table 1's note says "Dependant variable on [-1,1] scale". But the analysis
   file ships `corr_ideal_agency` already standardized to [0,1] -- `Data Cleaning
   II_Standardize_2014-04-08.do` line 297 applies (x + 1) / 2. Regressing the
   shipped column gives coefficients exactly HALF the published ones, with
   identical SEs-to-coefficient ratios, so nothing looks wrong. Undo it first:
       dv = corr_ideal_agency * 2 - 1

2. THE VARIABLE NAMED `neoliberal` RUNS BACKWARDS FROM ITS NAME.
   `Data Cleaning I_Generate_2014-04-08.do` lines 95-102 multiply the four
   constituent items by -1 under the comment "Recode to make (+) values liberal
   and (-) values conservative". `neoliberal` is then their average (line 143).
   So HIGHER `neoliberal` = MORE LIBERAL = LESS neoliberal. Confirm it yourself:
   `neoliberal` correlates +0.41 with `ideo_str` (+3 = very liberal).
   Table 1's "Neoliberal" row is the coefficient on the FLIPPED index, which the
   original calls `neoliberal_cons` (`definitions_agency_ORIGINAL.do` line 3):
       neoliberal_cons = neoliberal * -1
   Regress on `neoliberal` unflipped and you get +0.182 where the paper prints
   -0.182 -- same magnitude, wrong sign, and the constant still matches, so the
   error is easy to miss. The same flip applies to the four constituents in
   models 1b/2b/3b.

3. THE DUPLICATE SUBMISSION IS DROPPED.
   The analysis file has 992 rows but only 991 distinct respondents: two rows are
   identical on every column except `id` (1115 and 1157) -- one submission
   recorded twice. See ../data/PROVENANCE.md, "One respondent's record is
   duplicated". The published models drop it: keeping it gives N = 954/672 where
   the paper reports 953/671. No surviving script performs the drop; it is
   inferred from the numbers. Dropping it is what reproduces all six models'
   N and R-squared simultaneously -- and of all 992 rows, only these two do.

4. INTEGER GRID / FLOAT EQUALITY (not needed for Table 1, but see
   replicate_figures.py and the originals). The Stata code tests float equality
   directly (`if condition == float(2.1)`, `(income_household * 18) < 7`).
   Standardized values are stored as 0.7999999..., so exact or threshold tests
   on the un-rounded scale misfire. `income_household * 18` lands on an exact
   integer for only 109 of 992 rows. Round onto the integer grid first.

--------------------------------------------------------------------------------
STANDARD ERRORS
--------------------------------------------------------------------------------
The original runs `svy: reg` after `svyset [pweight = post_weight]`. With only a
pweight declared (no strata, no PSUs), Stata's linearized variance is the HC0
sandwich scaled by n/(n-1). That is what `svy_se()` below computes; it matches
the published SEs on ~130 of ~132 cells. (statsmodels' HC1 scales by n/(n-k)
instead and is very slightly off on some cells.)

The "Adj-R 2" row of the published table is NOT adjusted R-squared -- it is plain
R-squared. `svy: regress` reports only R-squared. Every one of the six published
values matches this script's `m.rsquared` to four decimals; none matches the
true adjusted R-squared. The row label is a misnomer.
"""

import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy import stats

DATA = "../../data/economic_justice_v7_analysis-weighted.csv"

# Published Table 1, transcribed from the article (p. 22) for cell-by-cell checking.
# Keyed by model -> variable -> (coefficient, standard error, stars).
PUBLISHED = {
    "1a": {"ideo_str": (-0.166, 0.093, "*"), "party_str": (0.007, 0.071, ""),
           "religiosity": (-0.079, 0.045, "*"), "theology": (0.020, 0.060, ""),
           "nonwhite": (-0.098, 0.032, "***"), "female": (0.005, 0.031, ""),
           "gay_strict": (0.047, 0.031, ""), "income_household": (0.108, 0.055, "*"),
           "school_ivy": (0.159, 0.051, "***"), "education": (0.028, 0.058, ""),
           "age": (-0.003, 0.001, "**"), "urban": (0.018, 0.039, ""),
           "south": (0.073, 0.029, "**"), "neo_c": (-0.269, 0.091, "***"),
           "const": (0.643, 0.073, "***")},
    "2a": {"ideo_str": (-0.043, 0.062, ""), "religiosity": (-0.039, 0.043, ""),
           "nonwhite": (-0.066, 0.029, "**"), "income_household": (0.123, 0.051, "**"),
           "school_ivy": (0.125, 0.041, "***"), "age": (-0.003, 0.001, "***"),
           "south": (0.039, 0.027, ""), "neo_c": (-0.201, 0.077, "***"),
           "const": (0.635, 0.058, "***")},
    "3a": {"nonwhite": (-0.070, 0.031, "**"), "income_household": (0.119, 0.049, "**"),
           "school_ivy": (0.114, 0.042, "***"), "age": (-0.003, 0.001, "**"),
           "neo_c": (-0.182, 0.058, "***"), "const": (0.623, 0.058, "***")},
    "1b": {"ideo_str": (-0.137, 0.091, ""), "party_str": (0.047, 0.0073, ""),
           "religiosity": (-0.076, 0.046, "*"), "theology": (0.032, 0.060, ""),
           "nonwhite": (-0.092, 0.030, "***"), "female": (0.015, 0.023, ""),
           "gay_strict": (0.028, 0.031, ""), "income_household": (0.121, 0.052, "**"),
           "school_ivy": (0.162, 0.053, "***"), "education": (0.014, 0.050, ""),
           "age": (-0.003, 0.001, "**"), "urban": (0.024, 0.037, ""),
           "south": (0.073, 0.026, "***"), "profits_c": (0.011, 0.052, ""),
           "gov_c": (0.039, 0.057, ""), "help_c": (-0.097, 0.052, "*"),
           "diff_c": (-0.193, 0.056, "***"), "const": (0.611, 0.076, "***")},
    "2b": {"religiosity": (-0.033, 0.035, ""), "nonwhite": (-0.061, 0.026, "**"),
           "income_household": (0.133, 0.049, "***"), "school_ivy": (0.121, 0.042, "***"),
           "age": (-0.003, 0.001, "***"), "south": (0.038, 0.027, ""),
           "help_c": (-0.085, 0.045, "*"), "diff_c": (-0.121, 0.048, "**"),
           "const": (0.601, 0.056, "***")},
    "3b": {"nonwhite": (-0.065, 0.028, "**"), "income_household": (0.125, 0.049, "***"),
           "school_ivy": (0.120, 0.043, "***"), "age": (-0.003, 0.001, "***"),
           "help_c": (-0.085, 0.045, "*"), "diff_c": (-0.122, 0.047, "***"),
           "const": (0.610, 0.055, "***")},
}
PUBLISHED_N = {"1a": 671, "2a": 953, "3a": 953, "1b": 671, "2b": 953, "3b": 953}
PUBLISHED_R2 = {"1a": 0.1978, "2a": 0.1288, "3a": 0.1193,
                "1b": 0.2320, "2b": 0.1467, "3b": 0.1382}

# Model specifications. Suffix _c marks a sign-flipped ("_cons") variable -- see
# gotcha 2. Models a use the neoliberal index; models b use its four constituents.
# Models 2 and 3 are reduced forms that drop insignificant predictors.
_FULL = ["ideo_str", "party_str", "religiosity", "theology", "nonwhite", "female",
         "gay_strict", "income_household", "school_ivy", "education", "age",
         "urban", "south"]
MODELS = {
    "1a": _FULL + ["neo_c"],
    "2a": ["ideo_str", "religiosity", "nonwhite", "income_household", "school_ivy",
           "age", "south", "neo_c"],
    "3a": ["nonwhite", "income_household", "school_ivy", "age", "neo_c"],
    "1b": _FULL + ["profits_c", "gov_c", "help_c", "diff_c"],
    "2b": ["religiosity", "nonwhite", "income_household", "school_ivy", "age",
           "south", "help_c", "diff_c"],
    "3b": ["nonwhite", "income_household", "school_ivy", "age", "help_c", "diff_c"],
}


def load():
    """Load the analysis layer and build the variables the published models use."""
    d = pd.read_csv(DATA, low_memory=False)

    # Gotcha 1: undo the [0,1] standardization to recover the published [-1,1] DV.
    d["dv"] = d["corr_ideal_agency"] * 2 - 1

    # Age is the one predictor left unstandardized (Table 1 note). Widen it: a
    # narrow integer dtype overflows silently on any squared/interaction term.
    d["age"] = d["age"].astype("float64")

    # Gotcha 2: flip the neoliberal index and its four constituents so that
    # HIGHER = MORE neoliberal, which is the orientation Table 1 reports.
    for src, dst in [("neoliberal", "neo_c"), ("profits_benefit", "profits_c"),
                     ("gov_stayout", "gov_c"), ("help_self", "help_c"),
                     ("diff_incentives", "diff_c")]:
        d[dst] = -d[src]

    # Gotcha 3: drop the duplicated submission (id 1115 / 1157 twins). Locate it
    # by content rather than row number so this survives a re-sort of the file.
    twins = d[d.duplicated(subset=[c for c in d.columns if c != "id"], keep="first")]
    d = d.drop(index=twins.index)

    return d


def svy_se(fit, n):
    """Stata `svy: reg` linearized SE = HC0 sandwich scaled by n/(n-1)."""
    return fit.get_robustcov_results(cov_type="HC0").bse * np.sqrt(n / (n - 1))


def stars(p):
    """Table 1 convention: * 10%, ** 5%, *** 1%."""
    return "***" if p < 0.01 else "**" if p < 0.05 else "*" if p < 0.10 else ""


def rnd(x, d=3):
    """Round half-up at the printed precision (numpy/Python round half-to-even
    would misreport cells like 0.1185 that sit exactly on the boundary)."""
    return float(np.format_float_positional(x, precision=d, unique=False,
                                            fractional=True))


def run(d, name):
    """Fit one published model; return a tidy frame plus N and R-squared."""
    xs = MODELS[name]
    s = d[["dv", "post_weight"] + xs].dropna()          # listwise, as Stata does
    X = sm.add_constant(s[xs], has_constant="add")
    m = sm.WLS(s["dv"], X, weights=s["post_weight"]).fit()
    n = int(m.nobs)
    se = svy_se(m, n)

    rows = []
    for i, v in enumerate(X.columns):
        b = m.params.iloc[i]
        t = b / se[i]
        # svy df = (number of PSUs) - (number of strata) = n - 1 here.
        p = 2 * stats.t.sf(abs(t), n - 1)
        rows.append({"var": v, "coef": b, "se": se[i], "p": p, "stars": stars(p)})
    return pd.DataFrame(rows), n, m.rsquared


def main():
    d = load()
    print("Table 1 -- Bower-Bir (2021), Economia Politica 38(1):131-170")
    print("Replication vs published. DV = corr_ideal_agency rescaled to [-1,1].")
    print('"Neoliberal"/constituent rows are the sign-flipped (_cons) variables.\n')

    mismatches = []
    for name in ["1a", "2a", "3a", "1b", "2b", "3b"]:
        res, n, r2 = run(d, name)
        okn = "OK" if n == PUBLISHED_N[name] else "MISMATCH"
        okr = "OK" if rnd(r2, 4) == PUBLISHED_R2[name] else "MISMATCH"
        print(f"=== Model {name}   N={n} (published {PUBLISHED_N[name]}) {okn}"
              f"   R2={r2:.4f} (published 'Adj-R2' {PUBLISHED_R2[name]:.4f}) {okr}")
        print(f"    {'variable':<18}{'coef':>9}{'se':>9}  {'':<4}"
              f"{'pub coef':>9}{'pub se':>9}")
        for _, r in res.iterrows():
            pub = PUBLISHED[name].get(r["var"])
            if pub is None:
                continue
            pc, ps, pst = pub
            bad = []
            if rnd(r["coef"]) != pc:
                bad.append("coef")
            if rnd(r["se"]) != ps:
                bad.append("se")
            if r["stars"] != pst:
                bad.append("stars")
            flag = "" if not bad else "  <-- " + ",".join(bad)
            print(f"    {r['var']:<18}{rnd(r['coef']):>9.3f}{rnd(r['se']):>9.3f}"
                  f"  {r['stars']:<4}{pc:>9.3f}{ps:>9.4f}{flag}")
            if bad:
                mismatches.append((name, r["var"], bad, r["coef"], r["se"],
                                   r["stars"], pc, ps, pst))
        print()

    print("-" * 78)
    if not mismatches:
        print("All cells reproduce.")
    else:
        print("Cells that do not reproduce (see README 'Corrigendum'):")
        for name, v, bad, b, se, st, pc, ps, pst in mismatches:
            print(f"  Model {name:<3} {v:<18} {','.join(bad):<12} "
                  f"replication {b:.4f} ({se:.4f}){st or '-'}   "
                  f"published {pc:.3f} ({ps}){pst or '-'}")


if __name__ == "__main__":
    main()

File: earning-our-place-ecopol-2020/analysis/replicate_table1.R
# Replicate Table 1 of Bower-Bir (2021), "Earning our place, more or less."
# Economia Politica 38(1): 131-170. https://doi.org/10.1007/s40888-020-00201-9
#
# Base R only -- no packages. lm() with weights, plus a hand-rolled sandwich
# estimator, so this runs on a bare R install.
#   Rscript replicate_table1.R
#
# See replicate_table1.py for the full commentary. The four traps in brief:
#
#  1. The DV ships standardized to [0,1] (Data Cleaning II line 297 applies
#     (x+1)/2), but Table 1 reports it on [-1,1]. Undo it: dv = x * 2 - 1.
#     Skip this and every coefficient is exactly half the published value.
#
#  2. The variable named `neoliberal` runs BACKWARDS from its name. Data
#     Cleaning I lines 95-102 flip the four constituent items so that "(+)
#     values [are] liberal"; `neoliberal` is their average. So higher
#     `neoliberal` = LESS neoliberal. Table 1's "Neoliberal" row is the
#     coefficient on the flipped index (`neoliberal_cons` in the original,
#     definitions_agency_ORIGINAL.do line 3). Unflipped you get +0.182 where
#     the paper prints -0.182 -- and the constant still matches, so it is easy
#     to miss. Same flip for the four constituents in models 1b/2b/3b.
#
#  3. The duplicated submission (id 1115/1157, identical on all but `id`) is
#     dropped by the published models: keeping it gives N=954/672 vs the
#     published 953/671. See ../data/PROVENANCE.md.
#
#  4. Stata `svy: reg` with only a pweight = HC0 sandwich scaled by n/(n-1).
#     The published "Adj-R 2" row is plain R-squared, not adjusted.

DATA <- "../../data/economic_justice_v7_analysis-weighted.csv"

d <- read.csv(DATA, stringsAsFactors = FALSE)

# --- Trap 1: recover the published [-1,1] dependent variable -----------------
d$dv <- d$corr_ideal_agency * 2 - 1

# --- Trap 2: flip the neoliberal index and its constituents ------------------
# so that HIGHER = MORE neoliberal, the orientation Table 1 reports.
d$neo_c     <- -d$neoliberal
d$profits_c <- -d$profits_benefit
d$gov_c     <- -d$gov_stayout
d$help_c    <- -d$help_self
d$diff_c    <- -d$diff_incentives

# --- Trap 3: drop the duplicated submission ---------------------------------
# Identify it by content (every column except `id`), not by row position.
cmp <- d[, setdiff(names(d), "id")]
d <- d[!duplicated(cmp), ]

FULL <- c("ideo_str", "party_str", "religiosity", "theology", "nonwhite",
          "female", "gay_strict", "income_household", "school_ivy",
          "education", "age", "urban", "south")
models <- list(
  "1a" = c(FULL, "neo_c"),
  "2a" = c("ideo_str", "religiosity", "nonwhite", "income_household",
           "school_ivy", "age", "south", "neo_c"),
  "3a" = c("nonwhite", "income_household", "school_ivy", "age", "neo_c"),
  "1b" = c(FULL, "profits_c", "gov_c", "help_c", "diff_c"),
  "2b" = c("religiosity", "nonwhite", "income_household", "school_ivy", "age",
           "south", "help_c", "diff_c"),
  "3b" = c("nonwhite", "income_household", "school_ivy", "age", "help_c",
           "diff_c")
)
pub_n  <- c("1a" = 671, "2a" = 953, "3a" = 953, "1b" = 671, "2b" = 953, "3b" = 953)
pub_r2 <- c("1a" = 0.1978, "2a" = 0.1288, "3a" = 0.1193,
            "1b" = 0.2320, "2b" = 0.1467, "3b" = 0.1382)

# Stata `svy: reg` linearized SE: HC0 sandwich, scaled by n/(n-1).
# With pweights, lm(weights=w) gives the same point estimates as Stata's [pw=w];
# only the variance needs building by hand.
svy_se <- function(fit, X, w, resid) {
  n <- nrow(X)
  bread <- solve(t(X) %*% (X * w))            # (X' W X)^-1
  meat <- t(X) %*% (X * (w^2 * resid^2))      # X' W^2 e^2 X
  V <- bread %*% meat %*% bread
  sqrt(diag(V) * n / (n - 1))
}

stars <- function(p) {
  if (p < 0.01) "***" else if (p < 0.05) "**" else if (p < 0.10) "*" else ""
}

cat("Table 1 -- Bower-Bir (2021), Economia Politica 38(1):131-170\n")
cat("Base-R replication. DV = corr_ideal_agency rescaled to [-1,1].\n")
cat("'Neoliberal'/constituent rows are the sign-flipped (_cons) variables.\n\n")

for (nm in names(models)) {
  xs <- models[[nm]]
  s <- d[, c("dv", "post_weight", xs)]
  s <- s[complete.cases(s), ]                 # listwise deletion, as Stata does

  f <- as.formula(paste("dv ~", paste(xs, collapse = " + ")))
  fit <- lm(f, data = s, weights = s$post_weight)

  X <- model.matrix(fit)
  se <- svy_se(fit, X, s$post_weight, residuals(fit))
  n <- nrow(s)
  b <- coef(fit)
  p <- 2 * pt(abs(b / se), df = n - 1, lower.tail = FALSE)  # svy df = n - 1

  # Weighted R-squared, centred on the weighted mean -- this is the quantity the
  # published table prints in its (mislabelled) "Adj-R 2" row.
  yhat <- fitted(fit); w <- s$post_weight; y <- s$dv
  mw <- sum(w * y) / sum(w)
  r2 <- 1 - sum(w * (y - yhat)^2) / sum(w * (y - mw)^2)

  cat(sprintf("=== Model %s   N=%d (published %d) %s   R2=%.4f (published %.4f) %s\n",
              nm, n, pub_n[[nm]], ifelse(n == pub_n[[nm]], "OK", "MISMATCH"),
              r2, pub_r2[[nm]], ifelse(round(r2, 4) == pub_r2[[nm]], "OK", "MISMATCH")))
  for (i in seq_along(b)) {
    cat(sprintf("    %-18s %8.3f (%.3f) %s\n",
                names(b)[i], b[i], se[i], stars(p[i])))
  }
  cat("\n")
}

cat(paste(rep("-", 70), collapse = ""), "\n")
cat("Known published defects this script does NOT reproduce (it prints the\n")
cat("correct values; see README Corrigendum):\n")
cat("  Model 1b party_str SE  -- published (0.0073), correct (0.073)\n")
cat("  Model 1b female SE     -- published (0.023),  correct (0.027)\n")
cat("  Model 3b income stars  -- published ***,      correct ** (p=0.0104)\n")
